helion/_compiler/type_propagation.py

```diff
@@ -1747,7 +1755,18 @@ def visit_Dict(self, node: ast.Dict) -> TypeInfo:
         return DictType(element_types=element_types, origin=self.origin())
 
     def visit_Name(self, node: ast.Name) -> TypeInfo:
-        return self.scope.get(node.id)
+        type_info = self.scope.get(node.id)
+
+        # Only replace device functions when we're in device context (inside tile loops)
+        if (
+            isinstance(type_info, CallableType)
+            and type_info.value in _device_fn_replacements
```

Comment:
This will miss calling `builtins.print` since that will show up as visit_Attribute rather than visit_Name.  And more generally, not all replaceable functions will go through `visit_Name`.